### PR TITLE
Generator fixup

### LINKED ISSUE #20: Do not allow returns with arguments in generators
The following is accepted by the compiler, even though it is not valid in Python:

    def f():
        yield 1
        return 1

There should be a compile error if mixing return with an argument and yield within the same function.

**Update: only disallow this in Python 2 (see below).**

### LINKED ISSUE #498: Full yield support, including yield expressions
In python 3.X yield is a statement (currently supported) but also an expression.
You can do something like this:

```python
from typing import Iterator

def test_yield() -> Iterator[int]:
    value = 1
    try:
        value = yield value  # Here is the yield expression
        yield value
        yield value + 1
    except TypeError as e:
        print(e)

gen = test_yield()
x1 = next(gen)
x2 = gen.send(2)
print(x1)  # 1
print(x2)  # 2
gen.throw(TypeError, 'ERROR')
gen.close()
```

Right now, mypy complains in the expression:

    _program.py: In function "test_yield":
    _program.py, line 6: Parse error before "value"

And, also, the Iterator type need the functions send, throw and close, but I think this go in typing project:

    _program.py, line 14: Iterator[int] has no attribute "send" 
    _program.py, line 17: Iterator[int] has no attribute "throw" 
    _program.py, line 18: Iterator[int] has no attribute "close"

The reference is here: [https://docs.python.org/3.4/reference/expressions.html#yieldexpr](https://docs.python.org/3.4/reference/expressions.html#yieldexpr)

### LINKED ISSUE #633: Mandatory Iterator with "yield"
Is it intended that the use of `yield` mandates the return type `Iterator`?
For example, if I want the squares of numbers:

    def get_squares(n: int) -> Iterable[int]:
        return [i**2 for i in range(n)]

I cannot reimplement it as an iterator:

    def get_squares(n: int) -> Iterable[int]:
        for i in range(n):
            yield i**2

without getting `Iterator function return type expected for "yield"` (even if `Iterator` actually is a subclass of `Iterable`).

However, it works with `yield from`:

    def get_squares(n: int) -> Iterable[int]:
        yield from (i**2 for i in range(n))

What is the reasoning?

### LINKED ISSUE #695: mypy is confused by return values from generators
This is fine:
```
from typing import Iterator

def foo() -> Iterator[int]:
    yield 42

x = next(foo()) + 1000
print(x)
```
But now try this:
```
from typing import Iterator

def foo() -> Iterator[int]:
    yield 42
    return 'abc'
```
This complains about the `return` statement:
```
g.py, line 5: Incompatible return value type: expected typing.Iterator[builtins.int], got builtins.str
```
It's matching the generator return value (which is only accessible by a caller that uses `yield from`) to the declared return type (which is just the type of what you get by calling `foo()`.

### LINKED ISSUE #836: Traceback analysing asyncio "item = yield from queue.get()"
Analysing:
```python
import asyncio

@asyncio.coroutine
def test(self):
	queue = asyncio.Queue(16384)
	sample = yield from queue.get()
```
Result:
```
  File "/home/gjc/projects/mollytest/env/lib/python3.4/site-packages/mypy/checker.py", line 1693, in visit_yield_from_expr
    if result_instance.type.fullname() == "asyncio.futures.Future":
AttributeError: 'AnyType' object has no attribute 'type'
```

### PR DESCRIPTION
**Pull request by ddfisher**
This commit fixes up the handling of Generators in several ways:
- it removes YieldStmt and YieldFromStmt, because they unnecessarily
  duplicate ExpressionStmt of YieldExpr/YieldFromExpr
- it ensures functions are properly marked as is_generator when
  appropriate
- it fixes a parsing bug where empty yield expressions would cause a parse
  error
- it checks Generator functions for the appropriate type signature at
  the function level, rather than at the yield statement level
- it adds support for return statements in Generator functions

fixes #20, fixes #836, fixes #695, fixes #633, fixes #498

### CODE REVIEW

File: mypy/checker.py [hunk 1]
```diff
@@ -552,6 +602,16 @@ def check_func_def(self, defn: FuncItem, typ: CallableType, name: str) -> None:
                     self.fail(messages.RETURN_TYPE_CANNOT_BE_CONTRAVARIANT,
                          typ.ret_type)
 
+            # check that Generator functions have the appropriate return type
+            if defn.is_generator:
+                if not self.is_generator_return_type(typ.ret_type):
+                    self.fail(messages.INVALID_RETURN_TYPE_FOR_GENERATOR, typ)
+
+                # Python 2 generators aren't allowed to return values
+                if self.pyversion[0] == 2 and typ.ret_type.type.fullname() == 'typing.Generator':
```
Comment (ddfisher) [inline mypy/checker.py]: I wasn't sure how to write a unit test for this, but it should probably have one.

File: mypy/checker.py [hunk 2]
```diff
@@ -1868,27 +1859,53 @@ def visit_call_expr(self, e: CallExpr) -> Type:
         return self.expr_checker.visit_call_expr(e)
 
     def visit_yield_from_expr(self, e: YieldFromExpr) -> Type:
-        # result = self.expr_checker.visit_yield_from_expr(e)
-        result = self.accept(e.expr)
+        return_type = self.return_types[-1]
+        subexpr_type = self.accept(e.expr, return_type)
+        if isinstance(subexpr_type, Instance):
+            if subexpr_type.type.fullname() == 'asyncio.futures.Future':
```
Comment (ddfisher) [inline mypy/checker.py]: I left this alone for now because I don't fully understand the coroutine logic.

File: mypy/checker.py [hunk 3]
```diff
@@ -1868,27 +1859,53 @@ def visit_call_expr(self, e: CallExpr) -> Type:
...
+        return_type = self.return_types[-1]
+        expected_item_type = self.get_generator_yield_type(return_type)
+        if e.expr is None:
+            actual_item_type = Void()  # type: Type
         else:
-            # self.msg.yield_from_invalid_operand_type(e.expr, e)
-            self.msg.yield_from_invalid_operand_type(e.expr.accept(self), e)
-        return result
+            actual_item_type = self.accept(e.expr, expected_item_type)
+            if hasattr(actual_item_type, 'args') and cast(Instance, actual_item_type).args:
```
Comment (ddfisher) [inline mypy/checker.py]: It wasn't entirely clear to me what the reasoning behind this is, so I left it alone for now.

**Comment by gvanrossum** [comment]
About the Travis failures... It fails for Python 3.4 and 3.5 only, and is related to asyncio. Here is a repro that fails when run with your PR:
```
import asyncio
@asyncio.coroutine
def foo() -> asyncio.Future[None]:
        yield from asyncio.sleep(0.1)
```
And since Future isn't a base class of Generator, it fails your code, with this error:
```
asy.py: note: In function "factorial":
asy.py:3: error: The return type of a generator function should be "Generator" or one of its supertypes
```
The tests containing this pattern are in mypy/test/data/pythoneval-asyncio.test, and are skipped in 3.2 and 3.3, where asyncio doesn't exist (the test framework only runs pythoneval-asyncio.test in 3.4 and higher -- see mypy/test/testpythoneval.py:31).

So what to do? I can make this error go away by adding a check for 'asyncio.futures.Future' to is_generator_return_type(). That seems pretty unprincipled though -- and what if some other async library were to introduce a different Future type?

It is also a lie. The function foo() above does not return a Future! It really does return a generator. It doesn't have any of Future's methods, like cancel(), done() or add_done_callback().

Everywhere a test defines a coroutine with return type `Future[T]` it should really use `Generator[Future[Any], None, T]`. But we might as well say `Generator[Any, None, T]` since nobody except the event loop cares about the Future that is technically yielded whenever a coroutine yields. And send() is only called by the event loop, with a None parameter (in asyncio/tasks.py).

That's still unfortunately verbose though. Also, the stubs for asyncio need to be changed -- they currently define several coroutines that claim to return Futures, which are all lies: either those APIs aren't coroutines, or they don't return Futures -- asyncio is sometimes a bit vague about whether something is a coroutine or a Future, and callers aren't supposed to care much -- they are expected to call asyncio.ensure_future() if they need a Future (e.g. to add a callback) and they're not expected to depend on specific behavior of generators/coroutines -- only yield from them.

Apparently so far all this was put off till the future (!) by the code you are fixing that deals with yield-from -- that code was quietly accepting the lies.

Note that there's a further wrinkle -- Awaitable. This only exists in 3.5 and later (both in collections.abc and in typing). However, I think that we can put off Awaitable (maybe file an issue about it). There are no tests involving it yet, so it won't distract us further here.

**Comment by ddfisher** [comment]
Thanks for the detailed explanation!  I definitely agree that we shouldn't be satisfied with lies here.

For the verbosity, it would be nice if we could define a type alias `Coroutine[T] = Generator[Any, None, T]`, but it looks to me like type aliases can't currently expose type parameters?  If that's the case, I'm not sure there's much better option than accepting the verbosity for now.  (We could instead make Coroutine subclass of Generator, but then we'd need to do add some silly special casing, because we don't want to allow all Generator subclasses...)  If we just write `Generator[Any, None, T]` (as you suggested) it's not even that much longer.

I also agree about punting on Awaitable support for now -- that seems like the sort of thing we should implement after we have the rest of the language working and fast.  I filed an issue (#1139) for it.

**Comment by JukkaL** [comment]
It would be great if somebody would think through the whole asynchronous topic and how mypy should behave ideally in light of Python 3.5 and other recent developments. Most of the existing code related to `yield from` etc. is by external contributors and I haven't been very careful about how principled the code is, with the philosophy that it's easier to improve something that's a little wrong than to start implementing it from scratch.

**Comment by gvanrossum** [comment]
So my recommendation to unblock this PR is to just change the asyncio tests and stubs to use `Generator[Any, None, X]` instead of `Future[X]` (where X is usually None or some simple concrete type like int). In the future we can figure out a more compact spelling of that (maybe `Coroutine[X]`, but Coroutine is only defined in 3.5 and later; so let's postpone solving that).

File: mypy/checker.py [hunk 4]
```diff
@@ -1400,36 +1470,43 @@ def visit_return_stmt(self, s: ReturnStmt) -> Type:
...
-                        )
+                if isinstance(typ, AnyType):
+                    return None
+
+                if isinstance(return_type, Void):
+                    # FuncExpr (lambda) may have a Void return.
+                    # Function returning a value of type None may have a Void return.
+                    if isinstance(self.function_stack[-1], FuncExpr) or isinstance(typ, NoneTyp):
+                        return None
+
```
Comment (gvanrossum) [inline mypy/checker.py]: Please get rid of this blank line (it looks weird in context).

File: mypy/checker.py [hunk 5]
```diff
@@ -1400,36 +1470,43 @@ def visit_return_stmt(self, s: ReturnStmt) -> Type:
...
-                        self.check_subtype(
-                            typ, self.return_types[-1], s,
-                            messages.INCOMPATIBLE_RETURN_VALUE_TYPE
-                            + ": expected {}, got {}".format(self.return_types[-1], typ)
-                        )
+                if isinstance(typ, AnyType):
+                    return None
+
+                if isinstance(return_type, Void):
+                    # FuncExpr (lambda) may have a Void return.
```
Comment (gvanrossum) [inline mypy/checker.py]: Can this code actually be invoked from analyzing a lambda? Also, I guess "may" here and on the next line means "is allowed to" right?

File: mypy/checker.py [hunk 6]
```diff
@@ -1400,36 +1470,43 @@ def visit_return_stmt(self, s: ReturnStmt) -> Type:
...
+                        + ": expected {}, got {}".format(return_type, typ)
+                    )
             else:
-                # Return without a value. It's valid in a generator and coroutine function.
-                if (not self.function_stack[-1].is_generator and
-                        not self.function_stack[-1].is_coroutine):
-                    if (not isinstance(self.return_types[-1], Void) and
-                            self.typing_mode_full()):
-                        self.fail(messages.RETURN_VALUE_EXPECTED, s)
+                # return without a value
```
Comment (gvanrossum) [inline mypy/checker.py]: I prefer comments (esp. whole-line comments) to always be properly capitalized sentences ending with correct punctuation.

File: mypy/checker.py [hunk 7]
```diff
@@ -1400,36 +1470,43 @@ def visit_return_stmt(self, s: ReturnStmt) -> Type:
...
+                # empty returns are valid in coroutines
```
Comment (gvanrossum) [inline mypy/checker.py]: Honestly I think 'coroutine' isn't a great term here -- in Python 3.5 it takes on a new meaning (referring to an 'async def').

File: mypy/checker.py [hunk 8]
```diff
@@ -1868,27 +1862,44 @@ def visit_call_expr(self, e: CallExpr) -> Type:
...
+            iter_method_type = self.expr_checker.analyze_external_member_access(
+                '__iter__',
+                subexpr_type,
+                AnyType())
+
+            generic_generator_type = self.named_generic_type('typing.Generator',
+                                                             [AnyType(), AnyType(), AnyType()])
+            iter_type, _ = self.expr_checker.check_call(iter_method_type, [], [],
+                                                        context=generic_generator_type)
+        elif isinstance(subexpr_type, AnyType):
```
Comment (gvanrossum) [inline mypy/checker.py]: Why not check for AnyType first? It's a much cheaper test.

File: mypy/test/data/check-expressions.test [hunk 9]
```diff
@@ -1322,8 +1322,10 @@ class str: pass
 def f(x: int) -> None:
     x = yield f('')
     x = 1
+[builtins fixtures/for.py]
 [out]
 main: note: In function "f":
+main:1: error: The return type of a generator function should be "Generator" or one of its supertypes
```
Comment (gvanrossum) [inline mypy/test/data/check-expressions.test]: I would rather fix up f()'s declaration to define the right return type? (e.g. Iterator[int])

File: mypy/test/data/check-statements.test [hunk 10]
```diff
@@ -51,6 +51,27 @@ def g() -> None:
...
+    return  # E: Return value expected
+[out]
+main: note: In function "f":
+
+[case testReturnInIterator]
+from typing import Iterator
+def f() -> Iterator[int]:
+    yield 1
+    return "foo"
+[out]
```
Comment (gvanrossum) [inline mypy/test/data/check-statements.test]: I guess there are also cases where the return type is `Generator[..., ... None]` and there's a plain `return` in the body (no error) or a `return 42` (error)?

File: mypy/test/data/check-expressions.test (hunk 9 above)
Comment (gvanrossum) [inline mypy/test/data/check-expressions.test]: Unless you're aiming for two birds with one stone?

File: mypy/checker.py (hunk 4 above)
Comment (ddfisher) [inline mypy/checker.py]: Done.

File: mypy/checker.py (hunk 5 above)
Comment (ddfisher) [inline mypy/checker.py]: This code will be invoked on lambdas: they have implicit `ReturnStmts` added to their AST.  My understanding is that "may" means "is allowed to" here (this comment/code was there before and just changed indentation levels).  I fixed up the comment to be clearer.

File: mypy/checker.py (hunk 6 above)
Comment (ddfisher) [inline mypy/checker.py]: I'll go through my comments and fix them up to have this style.

File: mypy/checker.py (hunk 7 above)
Comment (ddfisher) [inline mypy/checker.py]: Agreed.  We can actually get rid of this block entirely, which is the last remaining usage of `is_coroutine`.  `is_coroutine` is somewhat misleadingly named.  It means something closer to `CO_ITERABLE_COROUTINE` introduced in PEP 492, but that's not something we need to pay attention to at the moment (and I think probably needs to be carefully rethought when we do), so I removed it from the codebase.

File: mypy/checker.py (hunk 8 above)
Comment (ddfisher) [inline mypy/checker.py]: No good reason.  I'll switch the order.

File: mypy/test/data/check-expressions.test (hunk 9 above)
Comment (ddfisher) [inline mypy/test/data/check-expressions.test]: As it stands, it's checking both at once.  A common idiom in the tests seems to be purposefully introducing type errors to check the error message.  Testing the Generator function return type enforcement is important, but I don't have much of an opinion on whether it should be done in this function or a separate one.

File: mypy/test/data/check-statements.test (hunk 10 above)
Comment (ddfisher) [inline mypy/test/data/check-statements.test]: Yeah, thanks for catching that! Added.

**Pull request merged**